### PR TITLE
Better websocket reconnection handling

### PR DESCRIPTION
**Pull request by si14**
Right now it's fairly easy to encounter a situation when a tab coming online wouldn't recognise that the connection can now be reestablished for a while. This PR cleans up reconnection logic, reenables tests, and makes sure we get online as robustly as possible.

### Change Type

- [x] `patch` — Bug fix

### Test Plan

1. Check that reconnection works as expected

- [x] End to end tests

### CODE REVIEW

**Comment by ds300** [comment]
Returning to the retro tab a little later and it took about 5-6 seconds to reconnect, will keep trying.

**Comment by mimecuvalo** [comment]
> Mime went offline for no apparent reason, and it wasn't reconnecting, and he saw the offline indicator.

this looks like it was my wifi having issues possibly - i would disregard

**Comment by si14** [comment]
@ds300

Found, fixed, and documented an issue in code comments: sockets can hang in CLOSING state for quite a while and trigger their `.onclose` handlers much later. New version guards against that, explicitly orphans sockets on manual close, and does what it needs to do immediately instead of waiting for the delayed `.onclose`. This should be a bit more robust.

Additionally, I now do an opportunistic connectivity check on `window.offline`. This should make the transition to offline state faster.

File: apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts [hunk 1]
```diff
@@ -170,66 +198,255 @@ export class ClientWebSocketAdapter implements TLPersistentClientSocket<TLRecord
...
 		}
 	}
 
-	userInteractionOccurred() {
-		if (Date.now() + this.maxInteractiveIntervalLength < this.nextScheduledRunTimestamp) {
-			this.clear()
-			this.run()
+	connected() {
+		debug('ReconnectManager.connected')
+		// this notification cold've been delayed, recheck synchronously
```
Comment (ds300) [inline apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts:439]:
```suggestion
		// this notification could've been delayed, recheck synchronously
```

**Review by ds300 — APPROVED** [review]
Excellent work. This seems to function beautifully now!

File: apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts [hunk 2]
```diff
@@ -7,136 +7,160 @@ import {
...
-			windowListen('keydown', () => {
-				// if the user pressed a key while we are offline, we should try to reconnect more
-				// often than every 5 mins!
-				if (this.connectionStatus !== 'online') {
-					this._reconnectTimeout.userInteractionOccurred()
-				}
-			})
-		)
-		this._reconnectTimeout.run()
+	// TODO: this constructor is an adhoc interface; maybe it should be in TLPersistentClientSocket?)
```
[the remark this answers is not in the document]
Comment (si14) [reply apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts:59]: got it, thank you!

File: apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts [hunk 3]
```diff
@@ -170,66 +198,255 @@ export class ClientWebSocketAdapter implements TLPersistentClientSocket<TLRecord
...
+				this.socketAdapter._closeSocket()
+			})
+		)
+
+		this.state = 'pendingAttempt'
+		this.intendedDelay = ACTIVE_MIN_DELAY
+		this.scheduleAttempt()
+	}
+
+	private subscribeToReconnectHints() {
```
[the remark this answers is not in the document]
Comment (si14) [reply apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts:271]: It was inlined, I extracted it to make it a bit more readable. It's a bunch of subscriptions that serve two different purposes, so I think this is slightly cleaner.

File: apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts [hunk 4]
```diff
@@ -170,66 +198,255 @@ export class ClientWebSocketAdapter implements TLPersistentClientSocket<TLRecord
...
+			)
+		}
 	}
 
-	runNow() {
-		this.cb()
+	private scheduleAttempt() {
+		assert(this.state === 'pendingAttempt')
+		debug('scheduling a connection attempt')
+		Promise.resolve(this.getUri()).then((uri) => {
```
[the remark this answers is not in the document]
Comment (si14) [reply apps/dotcom/src/utils/remote-sync/ClientWebSocketAdapter.ts:299]: for future context (discussed offline): this serves as an inline "spawn", avoiding making the entire function async

**Pull request merged**